constructor-manager-ui/src/constructor_manager_ui/widgets/dialog.py

```diff
@@ -0,0 +1,465 @@
+        self._worker_packages = api.check_packages(
+            self.package_name,
+            version=self.current_version,
+            plugins_url=self.plugins_url,
+        )
+        self._worker_packages.finished.connect(self._update_packages)
+        self._worker_packages.start()
+
+        self._worker_updates = api.check_updates(
+            self.package_name,
+            current_version=self.current_version,
+            build_string=self.build_string,
+            channels=self.channels,
+            dev=self.dev,
+        )
+        self._worker_updates.finished.connect(self._update_widget)
+        self._worker_updates.start()
+
+        self.show_checking_updates_message()
+
+    def _update_version(self, result):
+        self._handle_finished(result)
+        if result.get("exit_code") == 0:
+            data = result["data"]
+            self.current_version = data.get("version", "")
+
+            if self.current_version:
+                text = f"v{self.current_version}"
+            else:
+                text = "(Not installed!)"
+
+            self.current_version_open_button.setEnabled(bool(self.current_version))
+
+            self.current_version_label.setText(f"{self.package_name} {text}")
+            self.current_version_open_button.setVisible(True)
+
+        self._refresh_after_version()
+
+    def _update_packages(self, result):
+        self._handle_finished(result)
+        data = result["data"]
+        package_data = []
+        if isinstance(data, dict):
+            packages = data.get("packages", [])
+            for pkg in packages:
+                package_data.append(
+                    PackageData(
+                        pkg["name"],
+                        pkg["version"],
+                        pkg["source"],
+                        pkg["build_string"],
+                        pkg["is_plugin"],
+                    )
+                )
+
+        self.set_packages(package_data)
+        self.set_busy(False)
+        self.refresh_button.setVisible(True)
+
+    def _update_widget(self, result):
+        self._handle_finished(result)
+        if result.get("exit_code") == 0:
+            data = result["data"]
+            if data.get("update"):
+                self.show_update_available_message(data.get("latest_version"))
+            else:
+                self.show_up_to_date_message()
+
+            states = data.get("states", [])
+            if states:
+                print(states)
+
+    # Qt Overrides
+    def closeEvent(self, event: QCloseEvent) -> None:
+        """Override Qt method."""
+        if self._busy:
+            QMessageBox.warning(
+                self, "Busy", "Please wait until the current action finishes!"
+            )
+            event.ignore()
+            return
+
+        # while ConstructorManagerWorker._WORKERS:
+        #     worker = ConstructorManagerWorker._WORKERS.pop()
+        #     self._terminate_worker(worker)
```

Comment:
Should this be removed or uncommented?

Reply:
Removed, we do not want to force close in the middle of an update.